lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -50,399 +37,218 @@
+        self.charm.plugin_manager.put_plugin_config(
+            scope=Scope.UNIT,
+            label=self.secret_label,
+            typ="keys",
+            cleanup=list(keys.keys()),
         )
 
-    def _needs_upgrade(self, plugin: OpenSearchPlugin) -> bool:
-        """Returns true if plugin needs upgrade."""
-        plugin_version = plugin.version.split(".")
-        version = self._opensearch.version.split(".")
-        num_points = min(len(plugin_version), len(version))
-        return version[:num_points] != plugin_version[:num_points]
+        self.charm.keystore.add_entries(keys)
+        self.charm.keystore.reload()
 
-    def _remove_if_needed(self, plugin: OpenSearchPlugin) -> bool:
-        """If disabled, removes plugin configuration or sets it to other values."""
-        return False
 
-    def _remove_plugin(self, plugin: OpenSearchPlugin) -> bool:
-        """Remove a plugin without restarting the node."""
-        try:
-            self._opensearch.run_bin("plugin", f"remove {plugin.name}")
-        except OpenSearchCmdError as e:
-            if "not found" in str(e):
-                logger.info(f"Plugin {plugin.name} to be deleted, not found. Continuing...")
-                return False
-            raise OpenSearchPluginRemoveError(plugin.name)
-        return True
-
-    @property
-    def _installed_plugins(self) -> List[str]:
-        """List plugins."""
-        try:
-            return self._opensearch.run_bin("plugin", "list").split("\n")
-        except OpenSearchCmdError as e:
-            raise OpenSearchPluginError("Failed to list plugins: " + str(e))
+class OpenSearchPluginEvents(Object):
+    """Events handler for OpenSearch plugin events"""
+
+    def __init__(self, charm: "OpenSearchBaseCharm"):
+        super().__init__(charm, "plugins")
+        self.charm = charm
+        self.framework.observe(
+            self.charm.on[PeerRelationName].relation_changed, self._on_peer_relation_changed
+        )
+
+    def _on_peer_relation_changed(self, _):
+        """Handle plugin secret-related peer relation changes."""
+        # if this is a subcluster, all units must add plugin keys from secrets to their keystores
+        if not self.charm.opensearch_peer_cm.is_consumer(of="main"):
+            return
+
+        app_plugins = self.charm.state.app.plugin_config_info
+        unit_plugins = self.charm.state.unit.plugin_config_info
+        added, removed = diff(app_plugins.keys(), unit_plugins.keys())
+        for label in added:
+            plugin = app_plugins[label]
+            if not plugin.secret_id:
+                continue
+
+            # start locally tracking secret and write transferred keys to keystore
+            keys_to_add = json.loads(
+                self.charm.secrets.track_secret(plugin.secret_id, Scope.APP, label)
+                .get_content()
+                .get(label)
+            )
```

Comment:
We can handle more gracefully here.
```suggestion
            content = self.charm.secrets.track_secret(plugin.secret_id, Scope.APP, label).get_content()
            raw = content.get(label)
            if not raw:
                continue
                
            try:
                keys_to_add = json.loads(raw)
            except json.JSONDecodeError:
                logger.error("Invalid JSON in secret for label %s", label)
                continue

```

Reply:
Thanks! Added